### PR TITLE
Added multiple entity state widget options

### PR DESCRIPTION
**Pull request by klejejs**
This PR adds entity state widget option to choose text size for the widget #754 and choose separator between entity state and attribute #756

### CODE REVIEW

**Comment by JBassett** [comment]
@klejejs  Sorry, I made some changes to use room instead of preferences because it was more of a hack to store the widget info.  Once your resolve the conflicts I can review,

**Comment by klejejs** [comment]
@JBassett is there a way to check if the migration is correct even after reinstalling the app? I think I messed up the DB, so I reinstalled the app and I obviously didn't have any problems after that.

**Comment by JBassett** [comment]
What I've been doing to validate is build and run the debug build from master and log in.  Then I check my branch back out with the migration and run it.  That should cause your migration to run  and you can validate the results.

We should add tests to the migrations to make it even better https://developer.android.com/training/data-storage/room/migrating-db-versions#test.  If you don't want to add the functionality in your PR I understand and I can add it in the future.

**Comment by klejejs** [comment]
@JBassett I manually checked migrations and it should be working now. I also looked into testing them, but concluded that it is out of scope for this PR. I unfortunately am not skilled enough to properly implement it.

File: app/src/main/java/io/homeassistant/companion/android/database/widget/StaticWidgetEntity.kt [hunk 1]
```diff
@@ -13,5 +13,9 @@ data class StaticWidgetEntity(
     @ColumnInfo(name = "attribute_id")
     val attributeId: String?,
     @ColumnInfo(name = "label")
-    val label: String?
+    val label: String?,
+    @ColumnInfo(name = "text_size")
+    val textSize: Int?,
+    @ColumnInfo(name = "separator")
+    val separator: String?
```
Comment (JBassett) [inline app/src/main/java/io/homeassistant/companion/android/database/widget/StaticWidgetEntity.kt:20]: Based on your other code it looks like these are required and always have a value, why have them nullable?

File: app/src/main/java/io/homeassistant/companion/android/migrations/Migrations.kt [hunk 2]
```diff
@@ -14,7 +14,7 @@ class Migrations constructor(
         private const val TAG = "Migrations"
         private const val PREF_NAME = "migrations"
         private const val PREF_VERSION = "migration_version"
-        private const val LATEST_VERSION = 3
+        private const val LATEST_VERSION = 4
```
Comment (JBassett) [inline app/src/main/java/io/homeassistant/companion/android/migrations/Migrations.kt:17]: This version just happened to match the DB version, if you aren't adding a non database migration you don't need to increment.

File: app/src/main/java/io/homeassistant/companion/android/widgets/StaticWidget.kt [hunk 3]
```diff
@@ -79,9 +82,18 @@ class StaticWidget : AppWidgetProvider() {
             val entityId: String? = widget?.entityId
```
Comment (JBassett) [inline app/src/main/java/io/homeassistant/companion/android/widgets/StaticWidget.kt:82]: Maybe we should put the rest of this method inside a null check for `widget`?  Then not everything will be nullable here and make the code easier

File: app/src/main/java/io/homeassistant/companion/android/widgets/StaticWidgetConfigureActivity.kt [hunk 4]
```diff
@@ -163,6 +158,16 @@ class StaticWidgetConfigureActivity : Activity() {
                 label.text.toString()
             )
 
+            intent.putExtra(
+                StaticWidget.EXTRA_TEXT_SIZE,
+                textSize.text.toString()
```
Comment (JBassett) [inline app/src/main/java/io/homeassistant/companion/android/widgets/StaticWidgetConfigureActivity.kt:163]: Shouldn't this be a number? `Float` or `Int`?

File: app/src/main/java/io/homeassistant/companion/android/widgets/StaticWidget.kt [hunk 5]
```diff
@@ -158,7 +173,9 @@ class StaticWidget : AppWidgetProvider() {
                 appWidgetId,
                 entitySelection,
                 attributeSelection,
-                labelSelection
+                labelSelection,
+                textSizeSelection?.toInt() ?: 30,
```
Comment (JBassett) [inline app/src/main/java/io/homeassistant/companion/android/widgets/StaticWidget.kt:177]: Should probably use `toIntOrNull()` so we don't throw and exception if it's not a number somehow.

Comment (JBassett) [inline app/src/main/java/io/homeassistant/companion/android/widgets/StaticWidget.kt:177]: Should probably use `toIntOrNull()` so we don't throw and exception if it's not a number somehow.

File: app/src/main/java/io/homeassistant/companion/android/database/widget/StaticWidgetEntity.kt (hunk 1 above)
Comment (klejejs) [reply app/src/main/java/io/homeassistant/companion/android/database/widget/StaticWidgetEntity.kt]: Makes sense. I hadn't fully understood what question mark in Kotlin means. Fixed now.

File: app/src/main/java/io/homeassistant/companion/android/migrations/Migrations.kt (hunk 2 above)
Comment (klejejs) [reply app/src/main/java/io/homeassistant/companion/android/migrations/Migrations.kt]: Got it. Fixed.

File: app/src/main/java/io/homeassistant/companion/android/widgets/StaticWidget.kt (hunk 3 above)
Comment (klejejs) [reply app/src/main/java/io/homeassistant/companion/android/widgets/StaticWidget.kt]: Also makes sense. Fixed that.

File: app/src/main/java/io/homeassistant/companion/android/widgets/StaticWidgetConfigureActivity.kt (hunk 4 above)
Comment (klejejs) [reply app/src/main/java/io/homeassistant/companion/android/widgets/StaticWidgetConfigureActivity.kt:163]: It is a number and I changed it to float in DB, however we are using a text field for input. Yes, you can only input numbers as specified in xml but as it is text field, it has to be a String. It is later converted to float when saving to DB.

File: app/src/main/java/io/homeassistant/companion/android/widgets/StaticWidget.kt (hunk 5 above)
Comment (klejejs) [reply app/src/main/java/io/homeassistant/companion/android/widgets/StaticWidget.kt]: Fixed.

File: app/src/main/java/io/homeassistant/companion/android/widgets/StaticWidget.kt [hunk 6]
```diff
@@ -158,7 +173,9 @@ class StaticWidget : AppWidgetProvider() {
                 appWidgetId,
                 entitySelection,
                 attributeSelection,
-                labelSelection
+                labelSelection,
+                textSizeSelection?.toFloatOrNull() ?: 30F,
+                if (separatorSelection == "") " " else separatorSelection ?: " "
```
Comment (klejejs) [inline app/src/main/java/io/homeassistant/companion/android/widgets/StaticWidget.kt:178]: @JBassett I think if user never touches the text field, the default value is an empty string not null, so can I write like this?

**Comment by klejejs** [comment]
@JBassett Thank you for doing the code review. I know my code is not perfect, but I am learning to get better ;)

**Pull request merged**